File: libvpx/test/VpxUnitTests/blockd_unittest.cpp
/******************************************************************************
* Filename: blockd_unittest.cpp
*
* Description: Unit testes relating to macro blocks.
*
* Author:
* Aaron Clauson (aaron@sipsorcery.com)
*
* History:
* 11 Nov 2020	Aaron Clauson	Created, Dublin, Ireland.
*
* License: Public Domain (no warranty, use at own risk)
/******************************************************************************/

#include "pch.h"
#include "strutils.h"
#include "CppUnitTest.h"
#include "vp8/common/alloccommon.h"
#include "vp8/common/reconintra.h"
#include "vp8/decoder/onyxd_int.h"
#include "vp8/encoder/block.h"
#include "vpx/vp8cx.h"

#include <iostream>
#include <string>
#include <vector>

using namespace Microsoft::VisualStudio::CppUnitTestFramework;

namespace VpxUnitTests
{
  TEST_CLASS(blockd_unittest)
  {
  public:

    TEST_METHOD(InitialiseMacroBlockTest)
    {
      struct frame_buffers fb;
      int res = vp8_create_decoder_instances(&fb, NULL);
      Assert::AreEqual((int)VPX_CODEC_OK, res);

      res = vp8_alloc_frame_buffers(&fb.pbi[0]->common, 640, 480);
      Assert::AreEqual(0, res);

      Assert::IsTrue(fb.pbi[0]->mb.block[24].qcoeff != 0);
    }

    TEST_METHOD(MacroBlockLayoutTest)
    {
      MACROBLOCKD mb;
      vp8_init_intra_predictors();
      Assert::IsTrue(true);
    }
  };
}

File: test/SIPSorcery.VP8.UnitTest/blockd_unittest.cs
//-----------------------------------------------------------------------------
// Filename: blockd_unittest.cs
//
// Description: Unit tests for logic in blockd.cs.
//
// Author(s):
// Aaron Clauson (aaron@sipsorcery.com)
//
// History:
// 04 Nov 2020	Aaron Clauson	Created, Dublin, Ireland.
//
// License: 
// BSD 3-Clause "New" or "Revised" License, see included LICENSE.md file.
//-----------------------------------------------------------------------------

using Xunit;

namespace Vpx.Net.UnitTest
{
    public unsafe class blockd_unittest
    {
        /// <summary>
        /// Checks that setting properties on the b_mode_info struct
        /// behaves as a union.
        /// </summary>
        [Fact]
        public void CheckBModeInfoUnionTest()
        {
            b_mode_info b = new b_mode_info();

            b.as_mode = B_PREDICTION_MODE.B_RD_PRED;
            Assert.Equal(5U, b.mv.as_int);

            b.mv.as_int = 1;
            Assert.Equal(B_PREDICTION_MODE.B_TM_PRED, b.as_mode);
        }

        [Fact]
        public void InitialiseMacroBlockTest()
        {
            frame_buffers fb = new frame_buffers();
            VP8D_CONFIG config = new VP8D_CONFIG();
            var errRes = onyxd.vp8_create_decoder_instances(fb, config);

            Assert.Equal(vpx_codec_err_t.VPX_CODEC_OK, errRes);

            int res = alloccommon.vp8_alloc_frame_buffers(fb.pbi[0].common, 640, 480);

            Assert.Equal(0, res);
            Assert.NotNull(fb.pbi[0].mb.block[24].qcoeff);
            //Assert.NotEqual(0, fb.pbi[0].mb.block[24].qcoeff.get());
        }
    }
}
